meteor/server/api/playout/adlib.ts

```diff
@@ -411,6 +411,56 @@ export namespace ServerPlayoutAdLibAPI {
 		})
 	}
 
+	export function innerFindLastScriptedPieceOnLayer(
+		cache: CacheForPlayout,
+		sourceLayerId: string[],
+		customQuery?: MongoQuery<Piece>
+	) {
+		const span = profiler.startSpan('innerFindLastScriptedPieceOnLayer')
+
+		const playlist = cache.Playlist.doc
+		const rundownIds = getRundownIDsFromCache(cache)
+
+		if (!playlist.currentPartInstanceId) {
+			return
+		}
+
+		const currentPartInstance = cache.PartInstances.findOne(playlist.currentPartInstanceId)
+
+		if (!currentPartInstance) {
+			return
+		}
+
+		const query = {
+			...customQuery,
+			startRundownId: { $in: rundownIds },
+			sourceLayerId: { $in: sourceLayerId },
+		}
+
+		const pieces = Pieces.find(query).fetch()
+
+		const part = Parts.findOne(
+			{ _id: { $in: pieces.map((p) => p.startPartId) }, _rank: { $lte: currentPartInstance.part._rank } },
+			{ sort: { _rank: -1 } }
+		)
+
+		if (!part) {
+			return
+		}
+
+		if (span) span.end()
+
+		return pieces
+			.sort((a, b) => {
```

Comment:
it would be better to make this a `pieces.filter().sort()[0]`, to make the sort quicker